src/VisualStudio/Core/Def/TaskList/ExternalErrorDiagnosticUpdateSource.cs

```diff
@@ -139,385 +110,160 @@ public ImmutableArray<DiagnosticData> GetBuildErrors()
-    /// raise diagnostic updated events for that project.
-    /// This ensures that error list keeps getting refreshed while a build is in progress, as opposed to doing all the work
-    /// and a single refresh when the build completes.
-    /// </summary>
-    private ValueTask ReportPreviousProjectErrorsIfRequiredAsync(ProjectId projectId, InProgressState state)
+    private static RpcContracts.DiagnosticManagement.DiagnosticSeverity GetSeverity(CodeAnalysis.DiagnosticSeverity severity)
     {
-        if (state.TryGetLastProjectWithReportedErrors() is ProjectId lastProjectId &&
-            lastProjectId != projectId)
-        {
-            return SetLiveErrorsForProjectAsync(lastProjectId, state);
-        }
-
-        return default;
+        return severity switch
+        {
+            CodeAnalysis.DiagnosticSeverity.Hidden => RpcContracts.DiagnosticManagement.DiagnosticSeverity.Hint,
+            CodeAnalysis.DiagnosticSeverity.Info => RpcContracts.DiagnosticManagement.DiagnosticSeverity.Information,
+            CodeAnalysis.DiagnosticSeverity.Warning => RpcContracts.DiagnosticManagement.DiagnosticSeverity.Warning,
+            CodeAnalysis.DiagnosticSeverity.Error => RpcContracts.DiagnosticManagement.DiagnosticSeverity.Error,
+            _ => throw ExceptionUtilities.UnexpectedValue(severity),
+        };
     }
 
-    private async ValueTask SetLiveErrorsForProjectAsync(ProjectId projectId, InProgressState state)
+    private static RpcContracts.Utilities.Range GetRange(DiagnosticData diagnostic)
     {
-        var diagnostics = state.GetLiveErrorsForProject(projectId);
-        await SetLiveErrorsForProjectAsync(projectId, diagnostics, state.CancellationToken).ConfigureAwait(false);
-        state.MarkLiveErrorsReported(projectId);
+        // Caller always created DiagnosticData with unmapped information.
+        var startPosition = diagnostic.DataLocation.UnmappedFileSpan.StartLinePosition;
+        var endPosition = diagnostic.DataLocation.UnmappedFileSpan.EndLinePosition;
+        return new RpcContracts.Utilities.Range(startPosition.Line, startPosition.Character, endPosition.Line, endPosition.Character);
     }
 
-    private ValueTask SetLiveErrorsForProjectAsync(ProjectId projectId, ImmutableArray<DiagnosticData> diagnostics, CancellationToken cancellationToken)
+    /// <summary>
+    /// Creates or gets the existing <see cref="IDiagnosticManagerService"/>
+    /// It is important that this is created only once as the client will remove our errors
+    /// when the instance of the brokered service is disposed of.
+    /// 
+    /// Serial access to this is guaranteed as all calls run inside the <see cref="_taskQueue"/>
+    /// </summary>
+    private async Task<IDiagnosticManagerService> GetOrCreateDiagnosticManagerAsync(CancellationToken cancellationToken)
     {
-        if (_diagnosticService is DiagnosticAnalyzerService diagnosticAnalyzerService)
+        if (_diagnosticManagerService == null)
```

Comment:
threading worriesm e here.  

Note: can we just make thsi simple and acquire this exactly once and hold onto it for the lifetime of roslyn from that point on?  is there a reason we need to actually dispose this?

Reply:
discussed offline - this is only ever created once, and is held onto the for the lifetime of roslyn (as a MEF part acquired by the VS Workspace).

Will only be disposed of on shutdown.

Reply:
i'm o kwith this given the invariant about only being created within the work queue